### ISSUE TITLE
Logging hangs when trying to close a STDOUT appender under JRuby

### ISSUE THREAD

**Issue by bklang** [TwP/logging#235]
This is a regression I noticed from 2.2.2 to 2.3.0.

There is code that attempts to [prevent calling `#close` on STDIO appenders](https://github.com/TwP/logging/blob/master/lib/logging/appenders/io.rb#L51). In logging 2.2.0 this worked because the IO appender was a `#<IO:<STDOUT>>` object. Since the [refactor to fix up console appenders](#220), that object is now an `#<IO:fd 1>`, and so is bypassing the conditional.

Interestingly, [tests exist for this scenario](https://github.com/TwP/logging/blob/master/test/appenders/test_console.rb#L46), but as far as I can tell, these tests are not being executed as part of the `rake test` run.

The problem is reproducible like this:

```Ruby
require 'logging'
Logging::Repository.instance
appender = Logging.appenders.stdout
appender.close # hangs - never returns
```

This problem appears under JRuby, but not under CRuby.
Tested broken under JRuby 9.1.17.0, 9.2.20.1, and 9.3.3.0.
Tested working with CRuby 2.7.4 and 3.1.0.

**Comment by TwP** [comment]
Finally taking a look at this today.

The test you [referenced above](https://github.com/TwP/logging/blob/master/lib/logging/appenders/io.rb#L51) is still working as expected. We don't want to close the Ruby defined IO streams `STDOUT` / `STDERR` / `STDIN`.

The Logging gem opens new IO streams for stdout and stderr. These streams can be closed and re-opened without affecting `STDOUT` and `STDERR`. This little snippet of code on Ruby 2.7 verifies this is the case:

```ruby
io = IO.new(STDOUT.fileno)
io.closed?  #=> false
STDOUT.closed  #=> false

io.close
io.closed?  #=> true
STDOUT.closed  #=> false
```

I'll poke at this more a bit later (I need to get a local JRuby environment set up to do this). In the meantime @bklang, does that small ruby snippet produce the same results in JRuby, or does that also hang the JRuby interpreter?

**Comment by bklang** [comment]
Thanks for taking  a look at this Tim!  Indeed, this does hang:

```
irb(main):001:0> RUBY_VERSION
=> "2.6.8"
irb(main):002:0> JRUBY_VERSION
=> "9.3.3.0"
irb(main):003:0> io = IO.new(STDOUT.fileno)
=> #<IO:fd 1>
irb(main):004:0> io.closed?
=> false
irb(main):005:0> STDOUT.closed?
=> false
irb(main):006:0>
irb(main):007:0> io.close
# *crickets*
```

**Comment by TwP** [comment]
The problem here is that `io = IO.new(STDOUT.fileno)` creates a new IO instance, but the underlying system file descriptor is still shared between our new IO object and `STDOUT`. So closing our copy of the IO instance will close the underlying system file descriptor; this causes `STDOUT` to be closed when our IO instance is closed.

I can modify the code to check if JRuby is being used and just not close our copy of the IO instance. However, as soon as our IO instance is garbage collected, the underlying system file descriptor will be closed resulting in the same hanging behavior.

The most direct solution is to _not_ create our own IO instance and just use `STDOUT` / `STDERR` directly. I'll poke at this a little more, but this is weird behavior in JRuby.

**Comment by TwP** [comment]
Just released version 2.3.1 that incorporates this fix.
